libcontainer/intelrdt/intelrdt.go

```diff
@@ -744,16 +837,79 @@ func (m *intelRdtManager) Set(container *configs.Config) error {
 	return nil
 }
 
-func (raw *intelRdtData) join(id string) (string, error) {
-	path := filepath.Join(raw.root, id)
-	if err := os.MkdirAll(path, 0755); err != nil {
-		return "", NewLastCmdError(err)
+func switchGroup(m *intelRdtManager, GetIntelRdtPath func(string, string) (string, error)) error {
+	// Prepare pids to move to another group.
+	buffer, err := ioutil.ReadFile(filepath.Join(m.GetPath(), IntelRdtTasks))
+	if err != nil {
+		return err
 	}
+	scanner := bufio.NewScanner(bytes.NewReader(buffer))
+	scanner.Split(bufio.ScanWords)
+	var pids []int
 
-	if err := WriteIntelRdtTasks(path, raw.pid); err != nil {
-		return "", NewLastCmdError(err)
+	for scanner.Scan() {
+		pid, err := strconv.Atoi(scanner.Text())
+		if err != nil {
+			return err
+		}
+		pids = append(pids, pid)
+	}
+
+	// Remove group.
+	if err := os.RemoveAll(m.path); err != nil {
+		return err
+	}
+
+	switch m.groupType {
+	case MonitoringGroupType:
+		m.groupType = ControlGroupType
+	case ControlGroupType:
+		m.groupType = MonitoringGroupType
+	}
+
+	m.path, err = GetIntelRdtPath(m.id, m.groupType)
+	if err != nil {
+		return err
+	}
+
+	// Prepare new group.
+	if err := os.MkdirAll(m.path, 0755); err != nil {
+		return err
+	}
+
+	for _, pid := range pids {
+		if err := ioutil.WriteFile(filepath.Join(m.path, IntelRdtTasks), []byte(strconv.Itoa(pid)), 0700); err != nil {
```

Comment:
How about calling WriteIntelRdtTasks() directly? And if any error, we could also get detailed error message from file `/sys/fs/resctrl/info/last_cmd_status`:

```
        if err := WriteIntelRdtTasks(m.path, pid); err != nil {
                return NewLastCmdError(err)
        }
```